sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/StarJoinCostBasedReorderSuite.scala

```diff
@@ -329,7 +329,7 @@ class StarJoinCostBasedReorderSuite extends PlanTest with StatsEstimationTestBas
     //
     // Number of generated plans: 46 (vs. 82)
     val query =
-      d1.join(t3).join(t4).join(f1).join(d2).join(t5).join(t6).join(d3).join(t1).join(t2)
+      d1.join(t3).join(t4).join(f1).join(d3).join(d2).join(t5).join(t6).join(t1).join(t2)
```

Comment:
Also CC @ioana-delaney as author of this test, on the off chance you have a comment on whether this test modification is valid. The last few comments above explain the issue a bit.

Reply:
@ioana-delaney Another infomation is Scala 2.13 tends to optimization results as 

```
Join Inner, (d1_pk#5 = f1_fk1#0)
:- Join Inner, (f1_fk2#1 = d2_pk#8)
:  :- Join Inner, (f1_fk3#2 = d3_pk#11)
```

and Scala 2.12 tends to  optimization results as 

```
Join Inner, (f1_fk2#1 = d2_pk#8)
:- Join Inner, (d1_pk#5 = f1_fk1#0)
:  :- Join Inner, (f1_fk3#2 = d3_pk#11)
```

If the input is reordered to `d1.join(t3).join(d3).join(t5).join(t1).join(t2).join(d2).join(f1).join(t6).join(t4)` in Scala 2.12, the optimization results  will change to 
 
```
Join Inner, (d1_pk#5 = f1_fk1#0)
:- Join Inner, (f1_fk2#1 = d2_pk#8)
:  :- Join Inner, (f1_fk3#2 = d3_pk#11)
```

Reply:
what? the scala version changes the result of the join reorder rule? This is very weird as the join reorder is data-dependent and should be deterministic. also cc @wzhfy

Reply:
@cloud-fan Yep, in this case 2 candidates plan in level 4 has same `Cost`, From the code, we can see which candidate is generated first and which one we will choice. 

Related codes as follow:

https://github.com/apache/spark/blob/3092527f7557b64ff9a5bedadfac8bb2f189a9b4/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala#L209-L231

if `newJoinPlan` betterThan `existingPlan` will use `newJoinPlan` else use  `existingPlan` and Same cost candidate not trigger update.

In Scala 2.13 `HashMap` and `HashSet` has been rewritten,  and I found the iteration order of `oneSideCandidates` and `otherSideCandidates`  from `foundPlans` are changes through debug this case.

Reply:
@wzhfy If my analysis is not correct, please correct me, thx ~

Reply:
Looks like it's due to the map entries order becomes different in scala 2.13. We should probably use a linked hash map to make the order deterministic.

Reply:
@cloud-fan I'll try this suggestion and give you a feedback

Reply:
@cloud-fan It seems that using "LinkedHashMap" does not make the results uniform, still two possibilities optimization results. The level 0 `foundPlans` is produced by `items.zipWithIndex`, the `items` produced by 

https://github.com/apache/spark/blob/3092527f7557b64ff9a5bedadfac8bb2f189a9b4/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala#L78-L91

If `itemIndex` has different order with different input order, use `LinkedHashMap` should be possible to guarantee the same optimization result for the same input order, unlikely to guarantee the same optimization results for different input order, If  the `items` is ordered, the output may be fully guaranteed, but I haven't verified it yet but I haven't verified it yet.

Now I am counting the optimization results of full permutation of 10 tables involved this case as input, divided into

- Scala 2.12 use HashMap
- Scala 2.12 use LinkedHashMap
- Scala 2.13 use HashMap
- Scala 2.13 use LinkedHashMap

The distribution of optimization results will be fed back later.

Reply:
@cloud-fan @srowen  I verified the above tests, the candidate input set is `Seq(d1, t3, t4, f1, d2, t5, t6, d3, t1, t2).permutations`, total of 3628000 input in different orders.

We define original expected optimization result A as

```
  f1.join(d3, Inner, Some(nameToAttr("f1_fk3") === nameToAttr("d3_pk")))
   .join(d1, Inner, Some(nameToAttr("f1_fk1") === nameToAttr("d1_pk")))
   .join(d2, Inner, Some(nameToAttr("f1_fk2") === nameToAttr("d2_pk")))
   .join(t4.join(t3, Inner, Some(nameToAttr("t3_c2") === nameToAttr("t4_c2"))), Inner,
     Some(nameToAttr("d1_c2") === nameToAttr("t3_c1")))
   .join(t2.join(t1, Inner, Some(nameToAttr("t1_c2") === nameToAttr("t2_c2"))), Inner,
     Some(nameToAttr("d3_c2") === nameToAttr("t1_c1")))
   .join(t5.join(t6, Inner, Some(nameToAttr("t5_c2") === nameToAttr("t6_c2"))), Inner,
     Some(nameToAttr("d2_c2") === nameToAttr("t5_c1")))
```

and define the other one optimization result B as

```
 f1.join(d3, Inner, Some(nameToAttr("f1_fk3") === nameToAttr("d3_pk")))
   .join(d2, Inner, Some(nameToAttr("f1_fk2") === nameToAttr("d2_pk")))
   .join(d1, Inner, Some(nameToAttr("f1_fk1") === nameToAttr("d1_pk")))
   .join(t4.join(t3, Inner, Some(nameToAttr("t3_c2") === nameToAttr("t4_c2"))), Inner,
     Some(nameToAttr("d1_c2") === nameToAttr("t3_c1")))
   .join(t2.join(t1, Inner, Some(nameToAttr("t1_c2") === nameToAttr("t2_c2"))), Inner,
     Some(nameToAttr("d3_c2") === nameToAttr("t1_c1")))
   .join(t5.join(t6, Inner, Some(nameToAttr("t5_c2") === nameToAttr("t6_c2"))), Inner,
     Some(nameToAttr("d2_c2") === nameToAttr("t5_c1")))
```

Some test results are as follows:

- Scala 2.12 use HashMap: 1813600 results were candidate A, 1814400 results were candidate B

- Scala 2.12 use LinkedHashMap: 1814400 results were candidate A, 1813600 results were candidate B

I will feedback on the test results in Scala 2.13 later.